packages/node/src/fs.js

```diff
@@ -0,0 +1,148 @@
+/**
+ * Filesystem utilities
+ *
+ * @packageDocumentation
+ */
+
+import assert from 'node:assert'
+import nodeFs from 'node:fs'
+import path from 'node:path'
+import { fileURLToPath } from 'node:url'
+
+/**
+ * @import {PathLike} from 'node:fs'
+ * @import {WithFs} from './internal.js'
+ */
+
+/**
+ * Reads a JSON file
+ *
+ * @template [T=unknown] Default is `unknown`
+ * @param {string | URL} filepath
+ * @param {WithFs} opts
+ * @returns {Promise<T>} JSON data
+ * @internal
+ */
+export const readJsonFile = async (filepath, { fs = nodeFs } = {}) => {
+  if (filepath instanceof URL) {
+    filepath = fileURLToPath(filepath)
+  }
+  const json = await fs.promises.readFile(filepath)
+  return JSON.parse(`${json}`)
+}
+
+/**
+ * Returns `true` if the file at `filepath` is a real file or a symlink to a
+ * real file
+ *
+ * @param {PathLike} filepath Path to check
+ * @param {WithFs} [options] Options
+ * @returns {boolean}
+ */
+export const isFileSync = (filepath, { fs = nodeFs } = {}) => {
+  return !!fs.statSync(filepath, { throwIfNoEntry: false })?.isFile()
+}
+
+/**
+ * Returns `true` if the file at `filepath` is readable
+ *
+ * @param {PathLike} filepath
+ * @param {WithFs} [options]
+ * @returns {boolean}
+ */
+export const isReadablePathSync = (filepath, { fs = nodeFs } = {}) => {
+  try {
+    fs.accessSync(filepath, fs.constants.R_OK)
+    return true
+  } catch {
+    return false
+  }
+}
+
+/**
+ * Returns `true` if the file at `filepath` is executable (and readable)
+ *
+ * @param {PathLike} filepath Path to check
+ * @param {WithFs} [options] Options
+ * @returns {boolean}
+ */
+export const isExecutablePathSync = (filepath, { fs = nodeFs } = {}) => {
+  try {
+    fs.accessSync(filepath, fs.constants.X_OK & fs.constants.R_OK)
```

Comment:
is bitwise AND what we intend here

Reply:
> `mode` should be either the value fs.constants.F_OK or a mask consisting of the bitwise OR of any of fs.constants.R_OK, fs.constants.W_OK, and fs.constants.X_OK (e.g. fs.constants.W_OK | fs.constants.R_OK)

think this either means
- only bitwise OR supported (remaining bitwise ops unsupported)
- only bitwise OR encouraged (remaining bitwise ops discouraged)

```suggestion
    fs.accessSync(filepath, fs.constants.X_OK)
```

seems the safe bet

Reply:
I think you're half-right; it should be `|` instead of `&`, but "execute" does not imply "read". In most cases, this must be true to actually execute the file.